lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,421 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mfaParams := sessionCtx.MFAParams(ap.GetRequireMFAType())
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		sessionCtx.Database.GetProtocol(),
+		sessionCtx.DatabaseUser,
+		sessionCtx.DatabaseName,
+	)
+	err = sessionCtx.Checker.CheckAccess(
+		sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	return trace.Wrap(err)
+}
+
+// getRoundTripper makes an HTTP client with TLS config based on the given URI host.
+func (e *Engine) getRoundTripper(ctx context.Context, uri string) (http.RoundTripper, error) {
+	if rt, ok := e.RoundTrippers[uri]; ok {
+		return rt, nil
+	}
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	hostname, err := getURIHostname(uri)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// We need to set the ServerName here because the AWS endpoint service prefix is not known in advance,
+	// and the TLS config we got does not set it.
+	tlsConfig.ServerName = hostname
+
+	out, err := defaults.Transport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	out.TLSClientConfig = tlsConfig
+	e.RoundTrippers[uri] = out
+	return out, nil
+}
+
+// getService extracts the service ID from the request header X-Amz-Target.
+func (e *Engine) getService(req *http.Request) (string, error) {
+	// try to get the service from x-amz-target header.
+	target := req.Header.Get(libaws.AmzTargetHeader)
+	if target == "" {
+		return "", trace.BadParameter("missing %q header in http request", libaws.AmzTargetHeader)
+	}
+	service, err := parseDynamoDBServiceFromTarget(target)
+	return service, trace.Wrap(err)
+}
+
+// getTargetURI returns the target URI constructed from a given service.
+func (e *Engine) getTargetURI(service string) (string, error) {
+	uri := e.sessionCtx.Database.GetURI()
+	if !apiaws.IsAWSEndpoint(uri) {
+		// non-aws endpoints are used without modification. This is an unusual configuration but supports unforeseen use-cases.
+		return uri, nil
+	}
+	// when the database is created we ensure any AWS endpoint is configured to be a suffix missing only the service prefix.
+	prefix, err := endpointPrefixForService(service)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return prefix + uri, nil
+}
+
+// rewriteRequest rewrites the request to modify headers and the URL.
+func rewriteRequest(ctx context.Context, r *http.Request, uri string, body io.Reader) (*http.Request, error) {
+	reqCopy, err := http.NewRequestWithContext(ctx, r.Method, r.URL.String(), body)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	reqCopy.Header = r.Header.Clone()
+
+	// force https and set url and host header to match the database uri.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.Host = uri
+	reqCopy.URL.Host = uri
+
+	for key := range reqCopy.Header {
+		// Remove Content-Length header for SigV4 signing.
+		if http.CanonicalHeaderKey(key) == "Content-Length" {
+			reqCopy.Header.Del(key)
+		}
+	}
```

Comment:
nit: do we need this loop?

Reply:
actually I think we don't, thanks for point that out!